packages/klighd-core/src/bookmarks/bookmark-registry.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * KIELER - Kiel Integrated Environment for Layout Eclipse RichClient
+ *
+ * http://rtsys.informatik.uni-kiel.de/kieler
+ * 
+ * Copyright 2021 by
+ * + Kiel University
+ *   + Department of Computer Science
+ *     + Real-Time and Embedded Systems Group
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+import { injectable, inject } from "inversify";
+import { Action, ICommand } from "sprotty";
+import { Registry } from "../base/registry";
+import { DISymbol } from "../di.symbols";
+import { PreferencesRegistry } from "../preferences-registry";
+import { Bookmark, GoToBookmarkAction, GoToBookmarkCommand, SetInitialBookmark } from "./bookmark";
+
+/**
+ * A simple {@link Registry} that holds a list of all added Bookmarks
+ *
+ * Handles CreateBookmark and GoToBookmark actions
+ * 
+ */
+@injectable()
+export class BookmarkRegistry extends Registry {
+
+    @inject(DISymbol.PreferencesRegistry) private preferenceRegistry: PreferencesRegistry;
+
+    private _bookmarks: Bookmark[] = [];
+    private _initialBookmark?: Bookmark;
+    private count = 0;
+
+    handle(action: Action): void | Action | ICommand {
+        if (action.kind === GoToBookmarkAction.KIND) {
+            return new GoToBookmarkCommand(action as GoToBookmarkAction, this.preferenceRegistry.preferences.animateGoToBookmark)
+        } else if (action.kind === SetInitialBookmark.KIND) {
+            this._initialBookmark = (action as SetInitialBookmark).bookmark
+            this.addBookmark(this._initialBookmark)
+        }
+    }
+
+    addBookmark(bookmark: Bookmark): void {
+        bookmark.bookmarkIndex = this.count++;
+        this._bookmarks.push(bookmark)
+        this.notifyListeners();
+    }
+
+    deleteBookmark(bookmark: Bookmark): void {
+        const index = this._bookmarks.findIndex((value) => value === bookmark);
+        this._bookmarks.splice(index, 1)
+        this.notifyListeners();
+    }
+
+    updateBookmarkName(bookmark_index: number, new_name: string): void {
+        const bm = this._bookmarks.find((bm) => bm.bookmarkIndex === bookmark_index)
+        if (bm) {
+            if (new_name === "") {
+                bm.name = undefined
+            } else {
+                bm.name = new_name
+            }
+            this.notifyListeners();
+        }
+
+    }
```

Comment:
Adjusted in commit 56468f69 to have all changes now go through the dispatcher, not sure if the CreateBookmarkCommand is ideal this way. I couldn't find a reasonable way to dispatch an AddBookmarkAction from the Command

Reply:
I see. The command can inject the `TYPES.ActionDispatcher` to dispatch the action. Commands should properly be pure operations and not have side-effects :thinking: . However, I know why the commands exist and I don't know a better way as well... Is it possible to dispatch Create/Delete actions in the undo/redo methods of the command as well? It would make the command more complete and overall bullet proof.

Also, a little pro-tip: You can add [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) to the actions to avoid the necessary casting when they are handled. Instead of comparing the action kind in the "if statement", it would be compared in the type predicate.
The [actions for options](https://github.com/kieler/klighd-vscode/blob/main/packages/klighd-core/src/options/actions.ts) should be a good example for such type predicate.

Reply:
Changed CreateBookmarkCommand to have the ActionDispatcher injected and use that to dispactch an AddBookmarkAction, also implemented undo and redo. 

Type predicates for the Actions have been added and are now used